### DISCUSSION TITLE
SB16 PnP/Vibra

### DISCUSSION THREAD

**Discussion by jschwartzenberg** [dosemu2/dosemu2#1507] [Ideas]
Some time back I was testing different SB16 drivers. Notably also:
https://archive.org/download/SoundBlaster16Software1997

This driver doesn't detect a device in dosemu2. In DOSBOX-X it detects one with `sbtype = sb16vibra` (not with the default `sbtype = sb16`).

When starting the installer this is logged:
`LOG: PNP OS changed I/O read port to 0x273 (from 0x20B)`

Would it make sense to support this? The majority of the DOS from Creative installers are kind of rough in dosemu2 and this driver has more features.

**Comment by stsp** [comment]
Would be cool to support ISA PnP,
but requires quite a bit of work.

**Reply by jschwartzenberg** [reply]
Alright then I assume that won't be in place soon. At least DOSBox-X could be inspiration.

I tried the `install-win31` repo and I found that now that all the win31 issues are solved or can be worked around (I use `$_cpu_vm_dpmi="native"` for now) it installs Windows 3.1 with the correct video & mouse drivers without issues. I want to add an audio driver too, but I am yet to come up with the perfect solution that avoids interaction.

**Reply by jschwartzenberg** [reply]
I just ran into another driver that just installs without sputtering on dosemu2, but it doesn't work in Windows for some reason (maybe a bug?):
http://www.toughdev.com/content/2017/07/creative-sound-blaster-16-waveffects-ct4170-on-a-80386/

**Reply by jschwartzenberg** [reply]
Older driver that just installs & works:
https://archive.org/details/ashley_SB16

Not sure how come I never managed to find these before. Or maybe there's something about my setup that's different from usual.... I'll do some more testing these days...

**Reply by jschwartzenberg** [reply]
@stsp you mentioned storing prepared images. Did you have a particular spot in mind? I'm looking for a good spot to host this :)
[sb16win31.tar.gz](https://github.com/dosemu2/dosemu2/files/6766108/sb16win31.tar.gz)

It's the win31 bits extracted from the above image which is >200 MB in total.

**Reply by stsp** [reply]
I meant a github build artefact.
But only Andrew knows how to
work with them. :)

**Reply by jschwartzenberg** [reply]
Ah that makes sense, yeah I should be able to come up with a GitHub Runner plan that pushes this as an artifact based on the file from `archive.org`. That should be a one-off build :D

**Reply by andrewbird** [reply]
Be aware that job artefacts have dynamically generated URLs, so you might be better off creating github pages for a static url and uploading it directly.

**Reply by jschwartzenberg** [reply]
Yeah I think I'll do that then. I hope that this variant of installing win31 could end up being done in a CI fashion at some point. It would have detected quite some regressions.

**Reply by andrewbird** [reply]
That would be a good thing. I guess the real challenge is driving the installation and developing some test of successful completion/function.

**Reply by jschwartzenberg** [reply]
Ah reading up on that, it seems it would end up on https://dosemu2.github.io/ if to be part of the project. Is anybody against hosting the above tarball there? Just to be clear, I intend to let it be fetched automatically in trigger its installation after installing win31 in the script here: https://github.com/dosemu2/install-win31

The script is intended to work without user interaction and install win31 with all dosemu2 drivers (video & mouse are already working). Future steps would be also installing important updates (y2k fix, euro sign, vfw, wing, etc.) and installing localized versions based on the LANG environment variable.

**Reply by andrewbird** [reply]
It's not really for me to say, but I've tried to keep proprietary binaries out of the repo myself.

**Reply by stsp** [reply]
I think there is a way to push
the artifacts elsewhere. I've seen
people uploading the github artifacts
to launchpad for example.
It would be better if we find some
hosting for proprietary binaries
and ask github to upload there.

**Reply by andrewbird** [reply]
> I think there is a way to push
the artifacts elsewhere.

Yes there are other standard actions for publishing, for example to Amazon services.

**Reply by jschwartzenberg** [reply]
Yeah I also kind of avoided putting proprietary stuff here previously, which is also why the DOS installers were working the way they were. There was also the idea to host preinstalled DOS environment though, which would break that.

Maybe we could have a spot on archive.org (or Amazon?) specifically for dosemu2 suitable driver sets? That would include the Trident drivers & the SB16 drivers to start.

**Reply by stsp** [reply]
There IS an idea to host
pre-installed DOS, but it
doesn't mean it should be
hosted on github. It should
only be produced by github.

**Reply by jschwartzenberg** [reply]
Ah I had missed that last bit. Did you have a particular host in mind?

**Reply by stsp** [reply]
I have no idea.
But what I do think is that its
not going to be some arbitrary
hosting, like someone's google
drive or an ftp at home.
We should create a site, or better
team up with the existing site
(like bootdisk project). Then at
least that would be a kinda "deal":
we just provide the downloader
from the famous site, and do not
care about the rest.
But if we provide the downloader
that just points to someone's google
drive, we'll get the ridicules.

Of course the site can be just the
trivial html with URLs, but you need
the right domain name at least.
Of course we can still use dyndns
or alike, да что угодно можно наколхозить,
but teaming up with someone else looks
like the best option to me.

**Reply by stsp** [reply]
@ecm-pushbx can bttr-software.de host
DOSes and Win31?

**Reply by stsp** [reply]
Have you considered more "free"
DOSes like edr-dos?

**Reply by ecm-pushbx** [reply]
We don't want to host any proprietary software with unclear or forbidden copyrights. It appears that archive.org does host MS-DOS and MS Windows install media though.

**Reply by stsp** [reply]
How about hosting EDR-DOS
and other "more open" DOSes?
We can as well forget about MS-DOS.
Perhaps even _should_. :)

**Reply by stsp** [reply]
So I guess we need to do the following:
- Forget about pre-configured proprietary DOSes.
- Add support for "more open" DOSes, like EDR-DOS,
for which it might be easier to find the hosting.
Unless their license says otherwise.
- As for win31, create the package with
"guest add-ons" and oemsetup.inf for an
easy installation.
- Use @jschwartzenberg 's automatic win31
installer for CI only.

**Reply by jschwartzenberg** [reply]
> * As for win31, create the package with
>   "guest add-ons" and oemsetup.inf for an
>   easy installation.
> 
> * Use @jschwartzenberg 's automatic win31
>   installer for CI only.

The installer would use the same package and could be provided to end users too. The matter of this whole thread is where to host this driver package :)

**Reply by stsp** [reply]
I don't think users need that auto-ocr
installer. They can install win31 themselves
quite easily, whereas packaging the
installer of an illegal windows copy,
may be illegal by itself.

In contrast, the driver package should
be legal to host anywhere, as I don't
think the drivers within, are forbidden
from free redistribution (and if so -
we need to avoid such drivers).

The question "where" still stays, but
we first need to make sure its legal
to host at all, then see where.

**Reply by jschwartzenberg** [reply]
> I don't think users need that auto-ocr
> installer. They can install win31 themselves
> quite easily, whereas packaging the
> installer of an illegal windows copy,
> may be illegal by itself.

You should check the new repo: https://github.com/dosemu2/install-win31
It doesn't use OCR at all, but uses the mechanism in win31's setup to perform the installation automatically once triggered by the user. This now integrates the mouse & video drivers too. (There is a Python script that integrates win31 drivers into a win31 installer included.)

**Reply by stsp** [reply]
I would suggest getting rid of
preinstallwin31, and instead
having the user to run that
script instead of setup.exe in
a win31 distro he got himself.

Does the comment still applies
that it doesn't work with fdpp?

**Reply by stsp** [reply]
We can still use the whole thing
for CI though.

**Reply by jschwartzenberg** [reply]
> Does the comment still applies
> that it doesn't work with fdpp?

Not anymore. I've only been testing this with FDPP, it all works fine. I was thinking of fetching localized win31 versions from archive.org based on someone `LANG` environment variable.

**Reply by stsp** [reply]
I think we should write that in
the doc, not in the code. The
user will have to click on URL
himself. This is not difficult at
all, and for us a big relief, as
providing the fetcher of illegal
copies may be itself illegal.

**Reply by jschwartzenberg** [reply]
We should at least draw a clear line for the project. Now the idea is: printing URLs on screen is ok, but fetching things from them is not. It's possible to work with that.

**Reply by stsp** [reply]
You can even put the entire script
into the doc if its not too large.
Normal practice for wiki page.

**Reply by stsp** [reply]
> Now the idea is: printing URLs on screen is ok

For wiki its definitely fine.
Printing by the program is maybe
not so much.

**Reply by stsp** [reply]
> We should at least draw a clear line for the project.

We can't, its not as simple as me
saying "I allow this and that".
There may be problems with fedora
or debian, so the best strategy is to
stay safe than sorry.

**Reply by stsp** [reply]
And in fact, I am much more upset
by not having ms-dos installer scripts,
that win31. :) Win31 is never a priority,
but ms-dos is good to check bugs/problems
with.

**Reply by jschwartzenberg** [reply]
> > We should at least draw a clear line for the project.
> 
> We can't, its not as simple as me
> saying "I allow this and that".
> There may be problems with fedora
> or debian, so the best strategy is to
> stay safe than sorry.

But that seems fine, we draw the line such that it fits those projects.


> And in fact, I am much more upset
> by not having ms-dos installer scripts,
> that win31. :) Win31 is never a priority,
> but ms-dos is good to check bugs/problems
> with.

As for the automation, we could do the same thing as proposed here with win31 for CI. Just build & install `msexpand` and use the script we have. If MS-DOS 1.x/2.x would do, those are on GitHub ;)

As a general note, I think it would be good to get clarity on archive.org. I think they have legitimacy at least in the hosting and offering part, but it's unclear what kind of usage is possible for us.

**Reply by stsp** [reply]
Why do I need msexpand if now
you can just run win setup under
dosemu and get it to completion?

**Reply by jschwartzenberg** [reply]
That would be in case you'd want to test MS-DOS in CI (similarly to testing win31 in CI).

**Reply by stsp** [reply]
No, we don't need to test MS-DOS
_setup_ in CI. win31 setup is much better
to get tested.

**Reply by jschwartzenberg** [reply]
There's no MS-DOS setup to test with the `msexpand` case as it's Python that does the installation, but maybe we could still fetch and install it automatically in CI and then automate some actual tests with it?

**Reply by stsp** [reply]
Well, we don't need to _do_ msdos setup
under CI, because its already there
and testing.

**Reply by jschwartzenberg** [reply]
Ah clear, I wasn't aware of that.

**Reply by stsp** [reply]
So in case you can make it a winsetup.bat
or maybe patchwin31.bat to patch the
already installed win31 with your drivers,
then we can as well include that into
dosemu2 itself, no need for a separate
package.

**Reply by jschwartzenberg** [reply]
Yep, that should be fine. The driver patching is implemented in Python though, so the batch file would trigger this:
https://github.com/dosemu2/install-win31/blob/master/src/dosemu-patchwin31inst

**Reply by stsp** [reply]
Seems like you are parsing the
oemsetup and do installation by
hands. How about using windows
setup to install these drivers, and
so only to use only DOS bat file
+host tools like curl?

**Reply by jschwartzenberg** [reply]
The batch file that starts everything is here:
https://github.com/dosemu2/install-win31/blob/master/src/bat/inswin31.bat

First it calls out to Python to:
1. Fetch win31 + video & mouse drivers
2. Integrate both drivers into the win31 installer (independent script that I linked above)

Then it starts the installer pointing to a prepared `dosemu.shh`. This starts the win31 setup in a special automation mode to automatically pick the right drivers and do a complete installation without asking questions.

At the end it fixes the path as it would be after a reboot. On boot there's this batch file that sets up the path:
https://github.com/dosemu2/install-win31/blob/master/src/bat/dosrc.d/win2path.bat

So everything is automated. The Python code can be used independently too in case you'd do a manual installation, but the repo already provides the batch files to automate the installation. I guess the easiest would be for you to try it out to see how it works.

**Reply by stsp** [reply]
I meant to run setup of already
installed windows just to get the
drivers installed, rather than to
parse oemsetup.inf in python.

**Reply by jschwartzenberg** [reply]
It would be interesting to see if that could be automated too yeah. I've not come across this previously. I suspect there are tools that take over the role of Windows' setup and just copy things in place and modify `system.ini`/`win.ini`.

**Reply by stsp** [reply]
That would be good as will allow to
just include such thing in a main dosemu2
package, and move the rest to CI.

**Reply by stsp** [reply]
Note that there is a text-mode setup
even in alredy installed windows.
Its runable from DOS and maybe
you don't need any extra tools.

**Reply by jschwartzenberg** [reply]
Yeah I'm aware of that and I used it manually before. I'd have to experiment with it to see if it can do all the tricks we're imagining now.

**Reply by jschwartzenberg** [reply]
> Note that there is a text-mode setup
> even in alredy installed windows.
> Its runable from DOS and maybe
> you don't need any extra tools.

I checked this out, but I am afraid it won't work. As it appears to be a copy of the one from the floppies, it shows the same options for `/?` including the ability to do for example `/h:f:\win31\dosemu.shh` but it doesn't actually do anything.

So the only way that I see to do this would be:
1. install win31 clean
2. install win31 with dosemu drivers

Compare them  carefully and build scripting that can morph number one into number two hoping it would work for any installation.

**Reply by stsp** [reply]
Try win3.11, maybe there are some
setups that work better than others.
Also maybe its a dosemu bug so please
make sure win setup doesn't work properly
under other emulators.

**Reply by jschwartzenberg** [reply]
Sorry I was unclear, I think this is actually intended behavior. The flag does fully work for win31 when performing an installation. It just doesn't do anything when setup is started in the "win31 is already installed, so just changing drivers"-mode. With the installation mode the flag works fine and is what the https://github.com/dosemu2/install-win31 scripts rely on.

win3.11 you mean wfw3.11 or the special 3.11? Generally I think wfw3.11 is less optimal for dosemu2 usage as it only supports enhanced mode and some features currently don't work in that mode.

**Reply by stsp** [reply]
So do you mean only GUI setup can
deal with oemsetup.inf?

**Reply by jschwartzenberg** [reply]
No, both can, but you have to do everything by hand once win3x is installed, whereas with setup initiated to perform a fresh installation it can be fully automated through a supplied file (ending in `.shh`).

**Reply by stsp** [reply]
Any pointers to what is .ssh?
Anyway, since it can still use .inf,
can that be usable for extra drivers?

**Reply by jschwartzenberg** [reply]
> Any pointers to what is .ssh?

This article describes it: https://jeffpar.github.io/kbarchive/kb/083/Q83194/
Note that it's .s*hh* not .ssh.

It mainly facilitates that setup picks the correct drivers without the user having to select them.

**Reply by stsp** [reply]
But can we get the work done with
oemsetup.inf instead? This looks more
like a custom driver install than the
batch setup, so why not oemsetup.inf?

**Reply by jschwartzenberg** [reply]
`oemsetup.inf` just contains the driver and a name for it. This is already included with the drivers that are necessary. Having a tool that installs those drivers automatically on top of win31 is what I'm looking at. The tool written now parses `oemsetup.inf` and added the entries inside the `setup.inf` used by the installer and copies the relevant files. Then it specifies `dosemu.shh` so that Windows Setup picks those drivers when installing.

I would say any driver will have a `oemsetup.inf` and we could possibly write a tool that takes it and instead of integrating it with Windows Setup, integrates it with an installed win31. That might be more complex, but could always be tried. Is that what you're suggesting?

**Comment by stsp** [comment]
> Is that what you're suggesting?

This is what I thought setup.exe
actually does. When you select the
custom driver location in GUI and
point to oemsetup.inf, you get your
driver copied and system.ini updated.

**Reply by jschwartzenberg** [reply]
Yep, but doing it manually that way is not so convenient, but I don't know if it could be automated?

**Reply by stsp** [reply]
Doesn't tui setup do the same?

**Reply by jschwartzenberg** [reply]
Depends, during installation it can do it automated, afterwards only the manual method is available afaik.

**Reply by stsp** [reply]
But you can do it manually in TUI?
Maybe then we can feed it with needed
key pressed?

**Reply by jschwartzenberg** [reply]
Yes, manually is no problem. I wouldn't know how to feed it decently ;) Maybe you have ideas there though?

**Reply by stsp** [reply]
Probably we need a tool that would
feed the needed keypresses after a
certain delay. During that delay you
run setup. :)
Well, not sure.
Perhaps all you need on a dos side,
is a gnu sed to make the changes
in an ini files, according to oemsetup?

**Reply by jschwartzenberg** [reply]
> Perhaps all you need on a dos side,
> is a gnu sed to make the changes
> in an ini files, according to oemsetup?

Yeah that's similar to what I intended, but I didn't have the idea to look for `sed` for DOS. For that ideally we have a fully free set of GNU(-like) utilities. They could be compiled with DJGPP, similarly to comcom32.

**Reply by stsp** [reply]
No, they can be downloaded from
ibiblio IIRC? I suppose install-freedosuserspace
should do that.

**Reply by jschwartzenberg** [reply]
Yeah, but being able to build from source (and having them up to date) is better imo.

**Reply by stsp** [reply]
Would you like to build the rest
of freedos as well?
No, we build only the essential
components. comcom32 is the one.

**Reply by jschwartzenberg** [reply]
> Would you like to build the rest
> of freedos as well?

Yes! No more install scripts, but everything in the main Debian, gNewSense, Parabola repos, etc. :) I'd even like to build win31 if possible but they haven't released the sources yet ;)

**Reply by stsp** [reply]
These scripts are completely optional.
Nothing prevents us from entering the
above distros... besides djgpp.
So lets say you take it all backwards.

**Reply by jschwartzenberg** [reply]
> Nothing prevents us from entering the
> above distros... besides djgpp.

Isn't DJGPP required for comcom32? Together with FDPP, that's key to enter those distros. Without you will end up in Debian contrib at max and not in gNewSense and Parabola at all.

**Reply by stsp** [reply]
The question was whether you want
to build the rest of freedos. Its not buildable
with djgpp, but with openwatcom.
So currently we at least do not depend
on openwatcom, even if djgpp alone
creates enough problems.

**Reply by jschwartzenberg** [reply]
Yeah openwatcom actually is struggling with its license regarding some silly clause that you need to make the compiler sources available when you offer binaries made with them even when those binaries are not free themselves or something like that.

**Reply by stsp** [reply]
I do not care about openwatcom as
freedos is completely optional.
What we _should_ care, is how to
get rid of djgpp instead.
Which is why I am saying you get
this all backwards.

**Reply by jschwartzenberg** [reply]
License wise at least the path for DJGPP seems open (just technically rough). But you have any ideas on getting rid of it?

**Reply by stsp** [reply]
The only idea is to port comcom32 on
clang, making it comcom64.

**Reply by jschwartzenberg** [reply]
Unless compiled on 32-bit (I compiled 32-bit fdpp too ;)
So what's easier? Packaging DJGPP for distros or porting comcom32 to clang? :)

**Reply by andrewbird** [reply]
Some FreeDOS utilities (label, find, fc, sort, choice) are now buildable with gcc-ia16, but more help is required for the rest. It crossed my mind that getting the utilities cross compiled from git and Debian packaged would be useful (if only because it's so long between FreeDOS releases), but it's a large task and probably no-one from the FreeDOS community would have any interest in either using it or contributing. I've been working on it off and on for a while, and @tkchia has been adding Watcom compatible functions into the libi86 library as necessary once I've discovered the need in each utility.

**Reply by stsp** [reply]
> So what's easier? Packaging DJGPP for distros or porting comcom32 to clang? :)

Just noticed this old question...
So what's the answer? :)

**Comment by stsp** [comment]
Of course packaging is easier, which is
why we do that now. But we can't do that
for every distro, and they can't package
it for us because its too difficult.
Also having a frame-work for porting
djgpp-based progs to clang would be
extra cool. fdpp has a very limited scope
of porting only freedos, but for comcom32
something more generic can be written,
in theory.

**Comment by stsp** [comment]
Even gcc-ia16 doesn't help unless you
upstream it and make accessible with
`gcc -m16`.
If you ever do that, then feel free to package
most useful freedos tools into our lp repo
and probably we'll get rid of
install-freedosuserspace then.
But to me its not a very big deal, just a
nice-to-have thing.

**Comment by jschwartzenberg** [comment]
I just pushed a commit that triggers the sb16 driver installation after installing win31:
https://github.com/dosemu2/install-win31/tree/feature/sb16driver

Just bashing keys and selecting 'exit to DOS' at the end of the Windows installation does the trick.

This is still ugly, but it helps me to get things reinstalled quickly in a fresh state and might help someone else too.

**Reply by stsp** [reply]
Can't see where in the patch you
do anything like that.

**Reply by jschwartzenberg** [reply]
Line 5 of `src/bat/inswin31.bat` starts the driver installer.

**Reply by stsp** [reply]
`Just bashing keys and selecting 'exit to DOS' at the end of the Windows installation does the trick.`
And this?

**Reply by jschwartzenberg** [reply]
Is up to the end user. While many prompts are, but not all prompts are automatically skipped.

**Reply by stsp** [reply]
Hmm.
So doing that from host side was better? :)

**Reply by jschwartzenberg** [reply]
I think some prompts can still be avoided.

Essentially I have implemented both extremes now and I am still planning to revisit the other scripts so they make use of this automation. Automating the Windows installer itself and reducing the amount of input the scripts would need to pass in, especially avoiding any graphical parts, will increase stability significantly.

**Reply by jschwartzenberg** [reply]
I just tried this installer of this particular driver itself again. Interestingly enough it fully ran to completion now. Before that never worked, it would refuse to alter the boot files.
I should try the other SB16 installers too some time.

**Reply by stsp** [reply]
So you mean I implemented ISApnp
already?

**Reply by jschwartzenberg** [reply]
No that branch uses a different driver: https://github.com/jschwartzenberg/dos-download-scripts/tree/master/sb16w31

**Reply by stsp** [reply]
Was there a bug ticket?
